### PR TITLE
Downloading the playbooks via the api

### PR DESCRIPTION
**Pull request by anara123**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: XSUP-12808

## Description
The issue happens because we download the playbook from the custom bundle. Instead we will download the playbook via XSOAR API.
Which will download the playbook via the field scriptName and not script.

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/download/downloader.py [hunk 1]
```diff
@@ -220,14 +220,38 @@ def fetch_custom_content(self) -> bool:
...
+
+                    if not self.list_files and re.search(r'playbook-.*\.yml', member.name):
+                        # if playbook and not list-file flag, we should download the file via direct REST API
+                        # because there are props like scriptName, that playbook from custom content bundle don't contain
+
+                        file_yaml_object = yaml.load(string_to_write)
+                        playbook_id = file_yaml_object.get('id')
+                        playbook_name = file_yaml_object.get('name')
+                        if playbook_id and playbook_name:
+                            if playbook_name not in self.input_files:
```
Comment (dantavori) [inline demisto_sdk/commands/download/downloader.py:239]: when self.all_custom_content flag is True, self.input_files at this stage still doesn't contain the playbook name (only after self.handle_all_custom_content_flag()) so we can't rely on it. Let's make sure handle_all_custom_content_flag is called before this function.

File: demisto_sdk/commands/download/downloader.py [hunk 2]
```diff
@@ -220,14 +220,38 @@ def fetch_custom_content(self) -> bool:
...
+                            if self.client and self.client.api_client and self.client.api_client.configuration:
+                                self.client.api_client.configuration.temp_folder_path = self.custom_content_temp_dir
+                                demisto_client.generic_request_func(self.client, f'/playbook/{playbook_id}/yaml', 'GET', response_type='file')
+                                self.client.api_client.configuration.temp_folder_path = None
+
+                                is_playbook_downloaded = True
+
+                    if not is_playbook_downloaded:
+                        file = open(file_path, 'w')
                         file.write(string_to_write)
```
Comment (dantavori) [inline demisto_sdk/commands/download/downloader.py:252]: let's keep using `with` or call `close()`

File: demisto_sdk/commands/download/downloader.py [hunk 3]
```diff
@@ -220,14 +220,38 @@ def fetch_custom_content(self) -> bool:
...
-                        string_to_write = extracted_file.read().decode('utf-8')
+
+                extracted_file = tar.extractfile(member)
+                # File might empty
+                if extracted_file:
+                    string_to_write = extracted_file.read().decode('utf-8')
+
+                    is_playbook_downloaded = False
+
+                    if not self.list_files and re.search(r'playbook-.*\.yml', member.name):
```
Comment (dantavori) [inline demisto_sdk/commands/download/downloader.py:231]: can we move this if block to a separate method handle_playbook_download()?

File: demisto_sdk/tests/integration_tests/download_integration_test.py [hunk 4]
```diff
@@ -37,6 +51,7 @@ def test_integration_download_no_force(demisto_client, tmp_path):
     - Ensure no download has been made.
     - Ensure force msg is printed.
     """
+    # mocker.patch.object(demisto_client, 'generic_request_func', return_value={'status_code': 200})
```
Comment (dantavori) [inline demisto_sdk/tests/integration_tests/download_integration_test.py:54]:
```suggestion
```

File: demisto_sdk/tests/integration_tests/download_integration_test.py [hunk 5]
```diff
@@ -11,17 +11,31 @@
...
+def match_request_text(client, url, method, response_type='text'):
+    if url == '/content/bundle':
+        with open('demisto_sdk/tests/test_files/download_command/demisto_api_response', 'r') as f:
+            api_response = f.read()
+
+        return (api_response, None, None)
+    elif url.startswith('/playbook') and url.endswith('/yaml'):
+        filename = url.replace('/playbook/', '').replace('/yaml', '')
+        # with open(f'demisto_sdk/tests/test_files/download_command/playbook-{filename}.yml', 'w') as f2:
+        #     f2.write(string_to_write)
```
Comment (dantavori) [inline demisto_sdk/tests/integration_tests/download_integration_test.py:23]:
```suggestion
```

**Review by dantavori — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/download/downloader.py (hunk 2 above)
Comment (dorschw) [reply demisto_sdk/commands/download/downloader.py:252]: with significantly preferred

File: demisto_sdk/commands/download/downloader.py [hunk 6]
```diff
@@ -203,6 +203,31 @@ def handle_max_retry_error(self, e):
...
+        playbook_id = file_yaml_object.get('id')
+        playbook_name = file_yaml_object.get('name')
+        if playbook_id and playbook_name and (playbook_name in self.input_files or self.all_custom_content):
+            # download the playbook yaml in case playbook name appears in the input_files or --all-custom-content flag is true
+
+            # this will make sure that we save the downloaded files in the custom cotent temp dir
+            if self.client and self.client.api_client and self.client.api_client.configuration:
+                api_resp = demisto_client.generic_request_func(self.client, f'/playbook/{playbook_id}/yaml', 'GET')
+                status_code = api_resp[1]
+                if status_code < 200 and status_code >= 300:
```
Comment (dantavori) [inline demisto_sdk/commands/download/downloader.py:224]:
```suggestion
                if status_code < 200 or status_code >= 300:
```

**Review by dantavori — APPROVED** [review]

**Pull request merged**